src/NimBLERemoteCharacteristic.cpp

```diff
@@ -202,42 +196,106 @@ int NimBLERemoteCharacteristic::descriptorDiscCB(uint16_t conn_handle,
 }
 
 
+/**
+ * @brief callback from NimBLE when the next characteristic of the service is discovered.
+ */
+int NimBLERemoteCharacteristic::nextCharCB(uint16_t conn_handle,
+                                           const struct ble_gatt_error *error,
+                                           const struct ble_gatt_chr *chr, void *arg)
+{
+    int rc = error->status;
+    NIMBLE_LOGD(LOG_TAG, "Next Characteristic >> status: %d handle: %d",
+                         rc, (rc == 0) ? chr->val_handle : -1);
+
+    ble_task_data_t *pTaskData = (ble_task_data_t*)arg;
+    NimBLERemoteCharacteristic *pChar = (NimBLERemoteCharacteristic*)pTaskData->pATT;
+
+    if (pChar->getRemoteService()->getClient()->getConnId() != conn_handle) {
+        return 0;
+    }
+
+    if (rc == 0) {
+        pChar->m_endHandle = chr->def_handle - 1;
+        rc = BLE_HS_EDONE;
+    } else if (rc == BLE_HS_EDONE) {
+        pChar->m_endHandle = pChar->getRemoteService()->getEndHandle();
+    } else {
+        pTaskData->rc = rc;
+    }
+
+    xTaskNotifyGive(pTaskData->task);
+    return rc;
+}
+
+
+/**
+ * @brief Get the end handle of this characteristic by discovering the next characteristic definition.
+ */
+bool NimBLERemoteCharacteristic::getNextCharHandle() {
+    NIMBLE_LOGD(LOG_TAG, ">> getEndHandle() for: %s", getUUID().toString().c_str());
+
+    int rc = 0;
+    ble_task_data_t taskData = {this, xTaskGetCurrentTaskHandle(), 0, nullptr};
+
+    rc = ble_gattc_disc_all_chrs(getRemoteService()->getClient()->getConnId(),
+                                 m_handle,
+                                 getRemoteService()->getEndHandle(),
+                                 NimBLERemoteCharacteristic::nextCharCB,
+                                 &taskData);
+    if (rc != 0) {
+        NIMBLE_LOGE(LOG_TAG, "Error getting end handle %d", rc);
+        return false;
+    }
+
+    ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+
+    if (taskData.rc != 0) {
+        NIMBLE_LOGE(LOG_TAG, "Could not retrieve end handle");
+    }
+
+    NIMBLE_LOGD(LOG_TAG, "<< getEndHandle()");
+    return (taskData.rc == 0);
+}
+
+
 /**
  * @brief Populate the descriptors (if any) for this characteristic.
  * @param [in] the end handle of the characteristic, or the service, whichever comes first.
  */
 bool NimBLERemoteCharacteristic::retrieveDescriptors(const NimBLEUUID *uuid_filter) {
     NIMBLE_LOGD(LOG_TAG, ">> retrieveDescriptors() for characteristic: %s", getUUID().toString().c_str());
 
-    uint16_t endHandle = getRemoteService()->getEndHandle(this);
-    if(m_handle >= endHandle) {
-        return false;
+    if (m_endHandle == 0) {
+        if (!getNextCharHandle()) {
+            return false;
+        }
+    } else if (m_handle == m_endHandle) {
+        return true;
     }
 
-    int rc = 0;
     ble_task_data_t taskData = {this, xTaskGetCurrentTaskHandle(), 0, nullptr};
```

Comment:
I feel that there are unfinished tasks in `xTaskGetCurrentTaskHandle()`.

Just before this line
```c
    taskYIELD();
```
After adding, the Panic error no longer occurs.

However, I'm not sure if this fix is ​​correct, and it's possible that the problem lurks on the side of `getNextCharHandle()`, which is running just before.